packages/hardhat-core/src/internal/hardhat-network/provider/vm/dual.ts

```diff
@@ -59,6 +69,24 @@ export class DualModeAdapter implements VMAdapter {
       selectHardfork
     );
 
+    // if the fork node config doesn't specify a fork block number, then the
+    // ethereum-js VM used the latest block number from the forked chain. here
+    // we get that number from that VM so it can be passed into the rethnet VM;
+    // otherwise, the rethnet VM's determination of "latest" might be a later
+    // number, resulting in a different fork block.
+    if (
+      isForkedNodeConfig(config) &&
+      config.forkConfig.blockNumber === undefined
+    ) {
+      const forkBlockNumber = ethereumJSAdapter.getForkBlockNumber();
+      if (forkBlockNumber !== undefined) {
```

Comment:
Can this happen? Shouldn't this be an assertion error?

Reply:
I instead did `forkBlockNumber`. I'll trust you that it's impossible within the control flow of the hardhat fork code 😅